Multi component repl

robrix opened the pull request:
This PR adds `script/repl`, which loads everything a `ghci` session with access to all of the packages’ sources, making it much easier to test stuff out without having to constantly quit & reload &c.

- [x] Depends on #369.

robrix commented on .ghci:
```diff
@@ -1,31 +1,28 @@
+-- :set -flocal-ghci-history
+
+-- -- Better typed holes
+-- :set -funclutter-valid-hole-fits -fabstract-refinement-hole-fits -frefinement-level-hole-fits=2
```
We should be defining these in our personal .ghci files because if we define them here then it can get super confusing as to what’s being defined where and what’s overriding what. Pretty-printing especially belongs out of here, so that we don’t end up loading multiple versions of the pretty-printing package and throwing errors on every reload. (Ask me how I know about that one.)

patrickt replied:
I remember that one. That one bit me during a demo. 😅

patrickt replied:
Since this file is now one big comment, can we uncomment the relevant `:set` calls and move this into a `.ghci.sample` file, so we can `cp .ghci.default ~/.ghc/.ghci`?

robrix replied:
That’s a superb idea, thank you! 👍

robrix commented on .ghci.semantic:
```diff
@@ -0,0 +1,70 @@
+:set -O0
+
+-- Compile to object code
+:set -fwrite-interface -fobject-code
```
I’m being a bit opinionated in this file:

1. I think we want to be able to quit the repl and relaunch it quickly.
2. I think we don’t want it to load all of the code automatically, since we’re working in various projects.
3. I think we especially don’t want it to import all of the code automatically.

I think those are fair statements for working on a variety of tasks, but if your particular workflow is made arduous by this, it’s worth revisiting.

robrix commented on .ghci.semantic:
```diff
@@ -0,0 +1,70 @@
+:set -Wno-unsafe
+:set -Wno-star-is-type
+-- Bonus: silence “add these modules to your .cabal file” warnings for files we :load
+:set -Wno-missing-home-modules
```
This one hits you for every `:load`ed file if you don’t turn it off.

patrickt commented on script/repl:
```diff
@@ -0,0 +1,15 @@
+repl_builddir=dist-repl
+
+if [[ ! -d $repl_builddir ]]; then
+  echo "$repl_builddir does not exist, first run 'cabal repl --builddir=$replbuilddir', exit, and then re-run $0"
```
Can we do this automatically? I guess not, since the repl is an interactive process…

patrickt replied:
Also, you made a typo here: it should be `'cabal repl --builddir=$repl_builddir’`.

robrix replied:
Oops, good catch!

patrickt commented:
Are we supposed to be able to see the packages/modules defined in `semantic`? Ah, I guess not, since this is a wrapper around `cabal exec ghci`.

robrix commented:
> Are we supposed to be able to see the packages/modules defined in `semantic`? Ah, I guess not, since this is a wrapper around `cabal exec ghci`.

Not sure what you mean? It doesn’t load them by default (because there’s no good way to anticipate what all you’ll need, there may be module name collisions in e.g. the tests, and you probably don’t want _everything_ in scope), so you still need to `:load Some.Specific.Module` or `:load path/to/Some/Specific/File.hs`, but everything that depends on should auto-load after that.

robrix commented:
@patrickt: I addressed your suggestions & also rewrote the relevant section of 💡ProTip!.md, so this is now ready for another glance, modulo https://github.com/github/semantic/pull/373#issuecomment-548774077.
